cosmwasm/contract.ts

```diff
@@ -0,0 +1,433 @@
+        options.description = options.description || `Create reward pools for ${chainName} voting verifier and multisig`;
+    }
+
+    const messages = [
+        JSON.stringify({
+            create_pool: {
+                params: {
+                    epoch_duration: epochDuration,
+                    participation_threshold: parsedThreshold,
+                    rewards_per_epoch: rewardsPerEpoch,
+                },
+                pool_id: {
+                    chain_name: chainName,
+                    contract: votingVerifierAddress,
+                },
+            },
+        }),
+        JSON.stringify({
+            create_pool: {
+                params: {
+                    epoch_duration: epochDuration,
+                    participation_threshold: parsedThreshold,
+                    rewards_per_epoch: rewardsPerEpoch,
+                },
+                pool_id: {
+                    chain_name: chainName,
+                    contract: multisigAddress,
+                },
+            },
+        }),
+    ];
+
+    return executeContractMessage(client, config, options, 'Rewards', messages, fee);
+};
+
+const instantiateChainContracts = async (
+    client: ClientManager,
+    config: ConfigManager,
+    options: ContractCommandOptions,
+    _args?: string[],
+    fee?: string | StdFee,
+): Promise<void> => {
+    const { chainName, salt, gatewayCodeId, verifierCodeId, proverCodeId, admin } = options;
+
+    const coordinatorAddress = config.validateRequired(config.getContractConfig('Coordinator').address, 'Coordinator.address');
+
+    if (!admin) {
+        throw new Error('Admin address is required when instantiating chain contracts');
+    }
+
+    if (!salt) {
+        throw new Error('Salt is required when instantiating chain contracts');
+    }
+
+    const chainConfig = config.getChainConfig(chainName);
+    const multisigProverContractName = config.getMultisigProverContractForChainType(chainConfig.chainType);
+
+    const gatewayConfig = config.getGatewayContract(chainName);
+    const votingVerifierConfig = config.getVotingVerifierContract(chainName);
+    const multisigProverConfig = config.getMultisigProverContract(chainName);
+
+    if (options.fetchCodeId) {
+        const gatewayCode = gatewayCodeId || (await getCodeId(client, config, { ...options, contractName: GATEWAY_CONTRACT_NAME }));
+        const votingVerifierCode =
+            verifierCodeId || (await getCodeId(client, config, { ...options, contractName: VERIFIER_CONTRACT_NAME }));
+        const multisigProverCode =
+            proverCodeId || (await getCodeId(client, config, { ...options, contractName: multisigProverContractName }));
+        gatewayConfig.codeId = gatewayCode;
+        votingVerifierConfig.codeId = votingVerifierCode;
+        multisigProverConfig.codeId = multisigProverCode;
+    } else {
+        if (!gatewayConfig.codeId && !gatewayCodeId) {
```

Comment:
same validation comment. in general, if you find yourself nesting more than one level, the code is not very readable